src/TreeSitter/Importing.hs

```diff
@@ -1,116 +1,146 @@
-                withCursor (castPtr rootPtr) $ \ cursor ->
-                  Just <$> runM (runReader cursor (runReader bytestring import'))
-      Exc.bracket acquire release go)
-
-withCursor :: Ptr TSNode -> (Ptr Cursor -> IO a) -> IO a
-withCursor rootPtr action = allocaBytes sizeOfCursor $ \ cursor -> Exc.bracket_
-  (ts_tree_cursor_new_p rootPtr cursor)
-  (ts_tree_cursor_delete cursor)
-  (action cursor)
-
-instance Importing a => Importing [a] where
-  import' = push $ do
-    a <- import' @a
-    pure [a]
-
-
-instance (Importing a, Importing b) => Importing (a,b) where
-  import' = push $ do
-    a <- import' @a
-    step
-    b <- import' @b
-    pure (a, b)
+import           GHC.Generics
+import           TreeSitter.Cursor as TS
+import           TreeSitter.Language as TS
+import           TreeSitter.Node as TS
+import           TreeSitter.Parser as TS
+import           TreeSitter.Tree as TS
+
+parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Maybe t)
+parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
+  if treePtr == nullPtr then
+    pure Nothing
+  else
+    withRootNode treePtr $ \ rootPtr ->
+      withCursor (castPtr rootPtr) $ \ cursor ->
+        runMaybeC (runM (runReader cursor (runReader bytestring buildNode)))
+
+-- | Building is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
+--
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
+class Building a where
+  buildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  default buildNode :: (Alternative m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  buildNode = to <$> push (getFields >>= gbuildNode)
+
+  buildEmpty :: Alternative m => m a
+  buildEmpty = empty
+
+instance Building Text.Text where
+  buildNode = do
+    node <- peekNode
+    case node of
+      Just node' -> do
+        bytestring <- ask
+        let start = fromIntegral (nodeStartByte node')
+            end = fromIntegral (nodeEndByte node')
+        pure (decodeUtf8 (slice start end bytestring))
+      _ -> empty
 
+instance Building a => Building (Maybe a) where
+  buildNode = Just <$> buildNode
+  buildEmpty = pure Nothing
 
-instance Importing Text.Text where
-  import' = do
-    node <- peekNode
-    bytestring <- ask
-    let start = fromIntegral (nodeStartByte node)
-        end = fromIntegral (nodeEndByte node)
-    pure (decodeUtf8 (slice start end bytestring))
+instance (Building a, Building b) => Building (Either a b) where
+  buildNode = Left <$> buildNode <|> Right <$> buildNode
 
+instance Building a => Building [a] where
```

Comment:
#### Why we need both `Building` and `GBuilding`

We have two classes: `Building` and `GBuilding`. This is because generic behaviors are different than what we get non-generically, and in the case of `Text`, `Maybe`, `[]`, `Either`—we actually preference doing things non-generically. Why? Well, since `[]` is a sum, the generic behaviour for `:+:` would be invoked and expect that we’d have repetitions represented in the parse tree as right-nested singly-linked lists (ex., `(a (b (c (d…))))`) rather than as just consecutive sibling nodes (ex., `(a b c ...d)`, which is what our trees have). We want to match the latter.